Caching when downloading via URL

dimensionv opened the pull request:
When GIFView is requested to download the GIF images via URL, caching that image falls into the responsibility of the View. It doesn't, when the caller provides an InputStream, but if the caller only provides an URL, the GIFView should do the caching.
Therefore, I have created this PR, with a class called GIFCache that hooks into the loading process when loading from an URL, and downloads the GIF image into the app's cache area, and opens an InputStream from there, passing that on, as before was done with the URL-inputstream.

Gavras commented:
For reference this PR is concerning #1.

First, It looks good, thank you.

Although a little thing that is bugging me. The using app has no control of these files and cannot monitor them.
I'm talking about hypothetical situation when an application have its own cache files and wants to maintain the size of the cache directory. It can, of course, infer which are the GIFView files by simply saying: "If I don't know it than it must be from GIFView", but what if the app don't even know its own cache files (Unlikely? or Costly).

As I see it the simplest solution may be creating those files in a designated sub-directory in the cache directory and than we only need to inform the users that use this library (in the documentation) about using the cache in the first place and where we put all those files.

What do you think?

dimensionv commented:
Good point, indeed. I have made the appropriate changes.

dimensionv commented:
I have also updated the JavaDoc, added a section about the caching to the README.md file and made access to the getCacheSubDir() method possible (class was package-private before :) ), and more convenient, as the method is static, now.

Gavras commented:
I think you meant to make getCacheSubDir() static since the user have no access to GIFCache.
Also GIFCache isn't public so that doesn't help either.
I think you can make the method static and use the constant you created ( the path will be the same every time) and delegate the call to GIFView because I don't think the user need to bother with GIFCache  at all, he only need to know about GIFView.

dimensionv commented:
Delegating it to GIFView is a possibility. I considered it briefly but didn't know how you might feel about it. I will add it. However, as there might be calls to remove the cached stuff from places totally different than the actual View - GIFCache might be the right place, still. Such cleaning ups may happen from an Activity or even a Service, that is totally disconnected to the ones containing the GIFView component.

dimensionv commented:
There, delegation method now in GIFView. GIFCache should remain public in order to enable developers to access it directly for reasons explained above.

Gavras commented:
You're right about the locations of the calls to that method. However it is just to make it easy for the user to use this library because I think it is better for them to talk with the library through only one class if possible.

Thank you again for your contribution!
I'll merge this PR.

dimensionv commented:
Thank you for accepting the PR. :) That's what I love about open source libraries :)
